drivers/counter/counter_ll_stm32_rtc.c

```diff
@@ -0,0 +1,354 @@
+
+	return ticks;
+}
+
+static int rtc_stm32_set_alarm(struct device *dev, u8_t chan_id,
+				const struct counter_alarm_cfg *alarm_cfg)
+{
+	struct tm alarm_tm;
+	time_t alarm_val;
+	LL_RTC_AlarmTypeDef rtc_alarm;
+	struct rtc_stm32_data *data = DEV_DATA(dev);
+
+	u32_t now = rtc_stm32_read(dev);
+	u32_t ticks = alarm_cfg->ticks;
+
+	if (data->callback != NULL) {
+		LOG_DBG("Alarm busy\n");
+		return -EBUSY;
+	}
+
+
+	data->callback = alarm_cfg->callback;
+	data->user_data = alarm_cfg->user_data;
+	data->absolute = alarm_cfg->absolute;
+
+	if (!alarm_cfg->absolute) {
+		ticks += now;
+	}
+
+	LOG_DBG("Set Alarm: %d\n", ticks);
+
+	alarm_val = (time_t)(counter_ticks_to_us(dev, ticks) / USEC_PER_SEC);
+
+	gmtime_r(&alarm_val, &alarm_tm);
+
+	/* Apply ALARM_A */
+	rtc_alarm.AlarmTime.TimeFormat = LL_RTC_TIME_FORMAT_AM_OR_24;
+	rtc_alarm.AlarmTime.Hours = alarm_tm.tm_hour;
+	rtc_alarm.AlarmTime.Minutes = alarm_tm.tm_min;
+	rtc_alarm.AlarmTime.Seconds = alarm_tm.tm_sec;
+
+	rtc_alarm.AlarmMask = LL_RTC_ALMA_MASK_NONE;
+	rtc_alarm.AlarmDateWeekDaySel = LL_RTC_ALMA_DATEWEEKDAYSEL_DATE;
+	rtc_alarm.AlarmDateWeekDay = alarm_tm.tm_mday;
+
+	LL_RTC_DisableWriteProtection(RTC);
+	LL_RTC_ALMA_Disable(RTC);
+	LL_RTC_EnableWriteProtection(RTC);
+
+	if (LL_RTC_ALMA_Init(RTC, LL_RTC_FORMAT_BIN, &rtc_alarm) != SUCCESS) {
+		return -EIO;
+	}
+
+	LL_RTC_DisableWriteProtection(RTC);
+	LL_RTC_ALMA_Enable(RTC);
+	LL_RTC_ClearFlag_ALRA(RTC);
+	LL_RTC_EnableIT_ALRA(RTC);
+	LL_RTC_EnableWriteProtection(RTC);
+
+	return 0;
+}
+
+
+static int rtc_stm32_disable_alarm(struct device *dev, u8_t chan_id)
+{
+	LL_RTC_DisableWriteProtection(RTC);
+	LL_RTC_ClearFlag_ALRA(RTC);
+	LL_RTC_DisableIT_ALRA(RTC);
+	LL_RTC_ALMA_Disable(RTC);
+	LL_RTC_EnableWriteProtection(RTC);
+
+	DEV_DATA(dev)->callback = NULL;
+
+	return 0;
+}
+
+
+static u32_t rtc_stm32_get_pending_int(struct device *dev)
+{
+	return LL_RTC_IsActiveFlag_ALRA(RTC) != 0;
+}
+
+
+static u32_t rtc_stm32_get_top_value(struct device *dev)
+{
+	const struct counter_config_info *info = dev->config->config_info;
+
+	return info->max_top_value;
+}
+
+
+static int rtc_stm32_set_top_value(struct device *dev, u32_t ticks,
+				counter_top_callback_t callback,
+				void *user_data)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(ticks);
+	ARG_UNUSED(callback);
+	ARG_UNUSED(user_data);
+
+	return -ENOTSUP;
+}
+
+
+static u32_t rtc_stm32_get_max_relative_alarm(struct device *dev)
+{
+	const struct counter_config_info *info = dev->config->config_info;
+
+	return info->max_top_value;
+}
+
+
+void rtc_stm32_isr(void *arg)
+{
+	struct device *const dev = (struct device *)arg;
+	struct rtc_stm32_data *data = DEV_DATA(dev);
+
+	u32_t now = rtc_stm32_read(dev);
+
+	if (LL_RTC_IsActiveFlag_ALRA(RTC) != 0) {
+
+		LL_RTC_DisableWriteProtection(RTC);
+		LL_RTC_ClearFlag_ALRA(RTC);
+		LL_RTC_DisableIT_ALRA(RTC);
+		LL_RTC_ALMA_Disable(RTC);
+		LL_RTC_EnableWriteProtection(RTC);
+
+		if (data && data->callback) {
+			data->callback(dev, 0, now,
```

Comment:
clear `data->callback` before this call to free channel.

Reply:
i still don't see `data->callback` being cleared. If you call `counter_set_channel_alarm` from that callback you'll get `-EBUSY` error.